src/app/clusters/soil-measurement-server/soil-measurement-server.cpp

```diff
@@ -0,0 +1,151 @@
+
+MeasurementData gMeasurements[MATTER_DM_SOIL_MEASUREMENT_CLUSTER_SERVER_ENDPOINT_COUNT + CHIP_DEVICE_CONFIG_DYNAMIC_ENDPOINT_COUNT];
+
+Globals::Structs::MeasurementAccuracyRangeStruct::Type soilMoistureMeasurementLimitsAccuracyRange[] = {
+    { .rangeMin = 0, .rangeMax = 100, .percentMax = MakeOptional(static_cast<chip::Percent100ths>(10)) }
+};
+
+CHIP_ERROR SoilMeasurementAttrAccess::Init()
+{
+    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
+
+    // Initialize the soil moisture measurement limits to default values
+    for (auto & measurement : gMeasurements)
+    {
+        measurement.soilMoistureMeasurementLimits = {
+            .measurementType  = Globals::MeasurementTypeEnum::kSoilMoisture,
+            .measured         = true,
+            .minMeasuredValue = 0,
+            .maxMeasuredValue = 100,
+            .accuracyRanges   = DataModel::List<const Globals::Structs::MeasurementAccuracyRangeStruct::Type>(
+                soilMoistureMeasurementLimitsAccuracyRange)
+        };
+        measurement.soilMoistureMeasuredValue.SetNull();
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void SoilMeasurementAttrAccess::Shutdown()
+{
+    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+}
+
+CHIP_ERROR SoilMeasurementAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == SoilMeasurement::Id);
+
+    switch (aPath.mAttributeId)
+    {
+    case SoilMoistureMeasurementLimits::Id: {
+        MeasurementData * data = SoilMeasurementDataForEndpoint(aPath.mEndpointId);
+        return aEncoder.Encode(data->soilMoistureMeasurementLimits);
+    }
+    case SoilMoistureMeasuredValue::Id: {
+        MeasurementData * data = SoilMeasurementDataForEndpoint(aPath.mEndpointId);
+        return aEncoder.Encode(data->soilMoistureMeasuredValue);
+    }
+    case ClusterRevision::Id: {
+        return aEncoder.Encode(kClusterRevision);
+    }
+    default: {
+        break;
+    }
+    }
+    return CHIP_NO_ERROR;
+}
+
+MeasurementData * SoilMeasurementDataForEndpoint(EndpointId endpointId)
+{
+    auto index = emberAfGetClusterServerEndpointIndex(endpointId, app::Clusters::SoilMeasurement::Id,
+                                                      MATTER_DM_SOIL_MEASUREMENT_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (index == kEmberInvalidEndpointIndex)
+    {
+        return nullptr;
+    }
+
+    if (index >= MATTER_ARRAY_SIZE(gMeasurements))
+    {
+        ChipLogError(NotSpecified, "Internal error: invalid/unexpected energy measurement index.");
+        return nullptr;
+    }
+    return &gMeasurements[index];
+}
+
+// This function is intended for the application to set the soil measurement accuracy limits to the proper values during init.
+// Given the limits are fixed, it is not intended to be changes at runtime, hence why this function does not report the change.
```

Comment:
Except sometimes it does report the change?

Reply:
I have updated the comment